doc/design/clustered-etcd-alert-store/design.md

```diff
@@ -0,0 +1,150 @@
+# Clustered Etcd Alert Store
+
+Type: Design Document (and PR)
+
+Date: 2019-09-20
+
+Author: David C Wang <dcwangmit01@gmail.com>
+
+## Status Quo
+
+The current Alertmanager clustering implementation will allow multiple
+instances of Alertmanager to sync recently-updated notification log entries and
+alert silences via the gossip protocol.  The syncing of alert state is not yet
+implemented.
+
+Currently, Alertmanager alert state is entirely stored in process memory.  It
+is lost upon restart of Alertmanager.  This, in conjunction with the fact that
+Alertmanager alert state is not synced with clustered peers presents several
+problems:
+
+* Alert emitters must send alerts to every single Alertmanager instance in the
+  cluster.  If some Alertmanager instances receive alerts that other instances
+  do not, their state will drift apart.
+
+  * Typically, the alert emitter is Prometheus which uses service discovery to
+    locate all Alertmanager instances, and then sends alerts to each individual
+    target.  Custom alert emitters may not have that service discovery
+    capability and must be explicity configured.
```

Comment:
> Custom alert emitters may not have that service discovery capability and must be explicity configured.

I don't understand how discovering / configuring e.g. 3 instead of 1 alertmanager instance is more than a linear additional overhead.

Reply:
Kubernetes may move Alertmanager pods around a will as nodes disappear, chaoskube kills them, rolling updates occur, the cluster autoscaler kills pods to pack them on other nodes before downscaling, etc.  Each of these occurrences can move an Alertmanager instance to new Pod with a new IP.  Unless the number of Alertmanager replicas never changes AND the alert emitters know how to query an updated target list from a Kubernetes headless service, the configuration will have to be manually managed.  We're hoping to avoid administrator intervention.